server/datastore/mysql/policies.go

```diff
@@ -283,63 +286,151 @@ func (ds *Datastore) RecordPolicyQueryExecutions(ctx context.Context, host *flee
+
+	var ids []uint
+	err := sqlx.SelectContext(ctx, q, &ids, initialQuery, args...)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "retrieving policy ids")
+	}
+
+	if len(ids) == 0 {
+		return []*fleet.Policy{}, nil
+	}
+
+	args = []interface{}{} // reset args
+
 	counts := `
-    (select count(*) from policy_membership where policy_id=p.id and passes=true) as passing_host_count,
-    (select count(*) from policy_membership where policy_id=p.id and passes=false) as failing_host_count
+	(select count(*) from policy_membership where policy_id=p.id and passes=true) as passing_host_count,
+	(select count(*) from policy_membership where policy_id=p.id and passes=false) as failing_host_count
 `
 	if countsForTeamID != nil {
 		counts = `
-        (select count(*) from policy_membership pm inner join hosts h on pm.host_id = h.id where pm.policy_id=p.id and pm.passes=true and h.team_id = ?) as passing_host_count,
-        (select count(*) from policy_membership pm inner join hosts h on pm.host_id = h.id where pm.policy_id=p.id and pm.passes=false and h.team_id = ?) as failing_host_count
+		(select count(*) from policy_membership pm inner join hosts h on pm.host_id = h.id where pm.policy_id=p.id and pm.passes=true and h.team_id = ?) as passing_host_count,
+		(select count(*) from policy_membership pm inner join hosts h on pm.host_id = h.id where pm.policy_id=p.id and pm.passes=false and h.team_id = ?) as failing_host_count
 `
 		args = append(args, *countsForTeamID, *countsForTeamID)
 	}
 
-	teamWhere := "p.team_id is NULL"
-	if teamID != nil {
-		teamWhere = "p.team_id = ?"
-		args = append(args, *teamID)
+	query := fmt.Sprintf(`
+		SELECT `+policyCols+`,
+			COALESCE(u.name, '<deleted>') AS author_name,
+			COALESCE(u.email, '') AS author_email,
+			%s
+		FROM policies p
+		LEFT JOIN users u ON p.author_id = u.id
+		WHERE p.id IN (?)`, counts)
+
+	args = append(args, ids)
+
+	query, args, err = sqlx.In(query, args...)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "building query to get policies by ID")
 	}
 
+	// removing pagination options to avoid double pagination
+	opts.Page = 0
+	opts.PerPage = 0
+
+	query = appendListOptionsToSQL(query, &opts)
+
 	var policies []*fleet.Policy
-	err := sqlx.SelectContext(
-		ctx,
-		q,
-		&policies,
-		fmt.Sprintf(`SELECT `+policyCols+`,
-      COALESCE(u.name, '<deleted>') AS author_name,
-      COALESCE(u.email, '') AS author_email,
-      %s
-    FROM policies p
-    LEFT JOIN users u ON p.author_id = u.id
-    WHERE %s`, counts, teamWhere), args...,
-	)
+	err = sqlx.SelectContext(ctx, q, &policies, query, args...)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "listing policies")
 	}
 	return policies, nil
 }
 
+// CountPolicies returns the total number of team policies.
+// If teamID is nil, it returns the total number of global policies.
+func (ds *Datastore) CountPolicies(ctx context.Context, teamID *uint, opts fleet.ListOptions) (int, error) {
```

Comment:
You are only using `opts.MatchQuery` here. Let's have a `matchQuery string` argument instead of `fleet.ListOptions`.